streaming/src/test/scala/org/apache/spark/streaming/receiver/BlockGeneratorSuite.scala

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.receiver
+
+import scala.collection.mutable
+
+import org.scalatest.BeforeAndAfter
+import org.scalatest.Matchers._
+import org.scalatest.PrivateMethodTester._
+import org.scalatest.concurrent.Eventually._
+import org.scalatest.time.SpanSugar._
+
+import org.apache.spark.storage.StreamBlockId
+import org.apache.spark.{SparkException, SparkConf, SparkFunSuite}
+
+class BlockGeneratorSuite extends SparkFunSuite with BeforeAndAfter {
+
+  import BlockGeneratorSuite._
+
+  private val conf = new SparkConf().set("spark.streaming.blockInterval", "100ms")
+  @volatile private var blockGenerator: BlockGenerator = null
+
+  before {
+    if (blockGenerator != null) {
+      blockGenerator.stop()
+    }
+  }
+
+  test("block generation and callbacks") {
+    val listener = new TestBlockGeneratorListener
+    blockGenerator = new BlockGenerator(listener, 0, conf)
+    require(listener.onAddDataCalled === false)
+    require(listener.onGenerateBlockCalled === false)
+    require(listener.onPushBlockCalled === false)
+    assert(blockGenerator.isStopped() === false)
+
+    blockGenerator.start()
+    assert(listener.onAddDataCalled === false)
+    assert(listener.onGenerateBlockCalled === false)
+    assert(listener.onPushBlockCalled === false)
+    assert(blockGenerator.isStopped() === false)
+
+    // Verify data is added and callback is called
+    val data1 = 1 to 10
+    data1.foreach { blockGenerator.addData _ }
+    assert(listener.onAddDataCalled === false)
+    assert(listener.onGenerateBlockCalled === false)
```

Comment:
Can be but very unlikely as the block interval is 100 ms. Let me try putting in a manual clock there.